### PR TITLE
Rubberband V3 stretcher

### PR DESCRIPTION
**Pull request by Swiftb0y**
if we're compiling with rubberband v3 available, automatically use the finer engine. The latest commit builds and works with Rubberband v2 and v3. This is not really intended to be merged because we want the engine to be configurable in the preferences. Consider this PR a starting point for anyone that wants to work on prefence integration.

### CODE REVIEW

File: src/engine/bufferscalers/enginebufferscalerubberband.cpp
Comment (daschuer) [inline src/engine/bufferscalers/enginebufferscalerubberband.cpp:27]: that is really to much magic. A good old `if()` in the using code does work and is way more readable IMHO:
```
if (IsEngineV2(m_pRubberBand)) {
... code for engine 2
} 
```
Is this a use case for the not_null ptr?

File: src/engine/bufferscalers/enginebufferscalerubberband.cpp
Comment (ronso0) [inline src/engine/bufferscalers/enginebufferscalerubberband.cpp:73]: the bitbucket link above is dead and can be removed

Comment (Swiftb0y) [reply src/engine/bufferscalers/enginebufferscalerubberband.cpp:73]: Didn't want to remove it for legacy sake, but yes, thats why I added the sourcehut link.

File: src/engine/bufferscalers/enginebufferscalerubberband.cpp
Comment (Swiftb0y) [reply src/engine/bufferscalers/enginebufferscalerubberband.cpp:27]: right. KISS.

> Is this a use case for the not_null ptr?

kinda, though if I simplify the `IsEngineV2` and make it a private member function, it could even work without any args.

Comment (Swiftb0y) [reply src/engine/bufferscalers/enginebufferscalerubberband.cpp]: done. Thanks for highlighting this obvious simplification.

Comment (ronso0) [reply src/engine/bufferscalers/enginebufferscalerubberband.cpp]: ot, though related, so I rather ask here instead of on [Zulip: Rubberband 3.0](https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Rubberband.203.2E0):
regarding the preferences integration, I'll add `Rubberband R3 (experimental)` to the keylock engine combobox. This comes from the `KeylockEngine` enum in [enginebuffer.h](https://github.com/mixxxdj/mixxx/blob/d53e0b0a421f3f7f9fb9927944978b948db90eb4/src/engine/enginebuffer.h#L77-L81) (hidden by `#if (RUBBERBAND_API_MINOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)`).

If this item is selected and applied (CO changed) [`EngineBuffer::slotKeylockEngineChanged`](https://github.com/mixxxdj/mixxx/blob/d53e0b0a421f3f7f9fb9927944978b948db90eb4/src/engine/enginebuffer.cpp#L801) would call something like `EngineBufferScaleRubberBand::setEngineV3(bool)` to set an internal bool `isEngineV3` which in turn is used to set the RubberBand options.

does this sound reasonable?

Comment (Swiftb0y) [reply src/engine/bufferscalers/enginebufferscalerubberband.cpp]: I'm not sure to be honest. I don't know how it is done internally but if it works for soundtouch, it should work for rubberband too. I'd also get rid of the `(experimental)` because R3 is quite stable, we just need to document that it takes even more CPU.

Somewhere you'll probably need to call RubberbandEngineScaler::onSampleRateChanged to construct the appropriate engine.

**Comment by Swiftb0y** [comment]
I also forgot to mention that with the R3 engine, changing the pitch results in temporary XRuns. We need to investigate that.

File: src/engine/bufferscalers/enginebufferscalerubberband.cpp
Comment (ronso0) [reply src/engine/bufferscalers/enginebufferscalerubberband.cpp]: > I'd also get rid of the (experimental) because R3 is quite stable

ah okay, I noticed the `-beta` tag but it's considered stable for a few days now.

**Comment by Be-ing** [comment]
It would be nice to add a CMake warning for old Rubberband versions noting that a new feature will not be available.

**Comment by Swiftb0y** [comment]
Sure, I can look into that.

**Comment by Swiftb0y** [comment]
> It would be nice to add a CMake warning for old Rubberband versions noting that a new feature will not be available.

I tried a bunch of things but can't get anything reliable. I can use the version acquired via PkgConfig, but that will only use the system installation and not the one actually used when specifing alternate library and include dirs. 

I don't consider it very important as well...

**Comment by Swiftb0y** [comment]
> I also forgot to mention that with the R3 engine, changing the pitch results in temporary XRuns. We need to investigate that.

I didn't find much time to work on this. I'm pretty sure its an issue within rubberband, but I need to spend more time confirming that. 
For now, I consider this PR good enough to be merged. It won't be used much anyways because of the lack of Rubberband 3 availability.

File: src/engine/bufferscalers/enginebufferscalerubberband.cpp
Comment (ywwg) [inline src/engine/bufferscalers/enginebufferscalerubberband.cpp:21]: can we ifdef it based on the rubberband version?

**Comment by ywwg** [comment]
interestingly I only get underruns when the musical pitch changes. If I turn on keylock I can move around the pitch slider and it doesn't drop at all (5ms buffer). But if I click the up-down arrows for key matching, or if I change key with keylock off and adjust the pitch slider (so that the music pitch changes) I get dropouts on every adjustment

**Review by ywwg — CHANGES_REQUESTED** [review]
Can we print out a warning during compilation (or cmake configure time) if the user has rubberband 3 installed to say that the library is only experimentally supported right now? otherwise if we merge this as-is we will get a lot of redundant reports about the dropout issues.

**Comment by ronso0** [comment]
@ywwg 
> This is not really intended to be merged because we want the engine to be configurable in the preferences. Consider this PR a starting point for anyone that wants to work on prefence integration.

I built on top of this in #4855 and added the pref option. Please try that (select RBv2) and verify it performs as it does in `main`.
Given the xruns, I'll re-add the "experimental" disclaimer to the RBv3 option in the preferences until someone tackles the cleanup in src/engine/bufferscalers/enginebufferscalerubberband.cpp

File: src/engine/bufferscalers/enginebufferscalerubberband.cpp
Comment (Swiftb0y) [reply src/engine/bufferscalers/enginebufferscalerubberband.cpp:21]: That suggestion is not trivial to implement so I'd like to avoid it. 
https://github.com/mixxxdj/mixxx/pull/4853#issuecomment-1185082936

Comment (Swiftb0y) [reply src/engine/bufferscalers/enginebufferscalerubberband.cpp:21]: I was only able to verify that this is not present the newer rubberband version. So I didn't want to introduce regressions based on speculation.

**Review by ywwg — APPROVED** [review]

**Review by daschuer — APPROVED** [review]
LGTM, thank you.

File: src/engine/bufferscalers/enginebufferscalerubberband.cpp
Comment (daschuer) [inline src/engine/bufferscalers/enginebufferscalerubberband.cpp:25]:
```suggestion
#define RUBBERBANDV3 (RUBBERBAND_API_MAJOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
```

**Review by daschuer — CHANGES_REQUESTED** [review]

File: src/engine/bufferscalers/enginebufferscalerubberband.cpp
Comment (Swiftb0y) [reply src/engine/bufferscalers/enginebufferscalerubberband.cpp]: Thanks. Idk how that slipped through.

**Comment by Swiftb0y** [comment]
ping me before merging please so I can autosquash the fixup commits.

**Comment by Swiftb0y** [comment]
Since this already got two approvals, I went a ahead and squashed anyways. Merge?

**Comment by daschuer** [comment]
Yes, thank you.

**Comment by ronso0** [comment]
I don't really understand why this was merged without adding a v2/v3 switch first? This is a performance regression for all who have v3 installed. Did I miss something? Or maybe I'm overestimating that because v3 didn't land in any distro yet anyway, and users who install it manually should be aware of the sideeffects?

**Comment by Swiftb0y** [comment]
> This is a performance regression for all who have v3 installed.

Its unlikely someone will have v3 installed. 

The plan was to merge #4855 asap if I'm not mistaken.

**Comment by daschuer** [comment]
> I don't really understand why this was merged without adding a v2/v3 switch first?

@ronso0 Oh I must have miss-read your initial comment in #4855 that you wait for this to be merged. 

But anyway, I am just working on the final bit and than we can hopefully merge that as well to get around possible regressions.

**Comment by ywwg** [comment]
yeah normally we like to preserve behavior but having a quick followup seems fine here

**Comment by ronso0** [comment]
> @ronso0 Oh I must have miss-read your initial comment in #4855 that you wait for this to be merged.

Sorry that was unclear. I expected this to be polished, than I'd rebase my followup and we merge both quickly one after another.
Anyway, probably only few are affected so I'll finish mine (ours) asap.

**Pull request merged**
